### DISCUSSION TITLE
Form validation using efx expressions from the SDK.

### DISCUSSION THREAD

**Discussion by Tronquill** [OP-TED/eForms-SDK#893] [Q&A] [answered]
Hi, I have a problem with validating notice fields on a form using SDK.

Assumption:
Determining whether a field is required or not , is done by executing expressions, specified for the field from sdk:
1) If any of conditions from forbidden constraint evaluates to true, field is forbidden.
2) If the field is not forbidden, I check if the field is required. The field is required, if any of the mandatory conditions evaluates to true

```
{
    "id" : "BT-5421-Lot",
    "forbidden" : {
      "value" : false,
      "severity" : "ERROR",
      "constraints" : [ {
        "noticeTypes" : [ "1", "2", "3", "4", "5", "6", "38", "39", "40", "CEI", "T01", "T02", "X01", "X02" ],
        "value" : true,
        "severity" : "ERROR"
      }, {
        "noticeTypes" : [ "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34", "35", "36", "37" ],
        "condition" : "{ND-LotAwardWeightCriterionParameter} ${BT-541-Lot-WeightNumber is not present}",
        "value" : true,
        "severity" : "ERROR"
      } ]
    },
    "mandatory" : {
      "value" : false,
      "severity" : "ERROR",
      "constraints" : [ {
        "noticeTypes" : [ "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34", "35", "36", "37" ],
        "value" : true,
        "severity" : "ERROR"
      } ]
    },
```

For field BT-5421-Lot for notice 15 I try to determine whether the field is required or not.
for this purpose i performs the following steps:

1. generate XML based on the already filled form (nodes are created, only for fields with filled value)
2. convert efx expressions to xpath using efx-tool
3. for each expression I specify the context node, taking the first element from the efx expression  (for the expression  "{ND-LotAwardWeightCriterionParameter} ${BT-541-Lot-WeightNumber is not present}", context node is set for ND-LotAwardWeightCriterionParameter (i'm searching for the element using assigned xpath expression from xpathAbsolute 
property of ND-LotAwardWeightCriterionParameter)
4. execute xpath  expression  on context node, to find if field is forbidden ( for field BT-5421-Lot, exp:  not(efbc:ParameterNumeric))

**How to determine if a field is forbidden or mandatory, if context node is missing? I can't compute xpath expression in void.** 

I have tried to always return true in case of missing context node, but it doesn't work because sometimes xpath expression requires dependant node to be missing, and sometimes xpath requires dependant node to be present.

**Comment by YvesJo** [comment]
Hi,
Could you please share a situation where the condition would be missing the context?
Thanks

**Comment by Tronquill** [comment]
sample xml:

problems occur when trying to validate the GR-Procedure-TenderingTerms section.

`<?xml version="1.0" encoding="UTF-8"?>
<ContractNotice xmlns="urn:oasis:names:specification:ubl:schema:xsd:ContractNotice-2"
                xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2"
                xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2"
                xmlns:efac="http://data.europa.eu/p27/eforms-ubl-extension-aggregate-components/1"
                xmlns:efbc="http://data.europa.eu/p27/eforms-ubl-extension-basic-components/1"
                xmlns:efext="http://data.europa.eu/p27/eforms-ubl-extensions/1"
                xmlns:ext="urn:oasis:names:specification:ubl:schema:xsd:CommonExtensionComponents-2">
    <ext:UBLExtensions>
        <ext:UBLExtension>
            <ext:ExtensionContent>
                <efext:EformsExtension>
                    <efac:NoticeSubType>
                        <cbc:SubTypeCode listName="notice-subtype">15</cbc:SubTypeCode>
                    </efac:NoticeSubType>
                    <efac:Organizations>
                        <efac:Organization>
                            <efac:Company>
                                <cac:PartyIdentification>
                                    <cbc:ID schemeName="organization">ORG-0001</cbc:ID>
                                </cac:PartyIdentification>
                            </efac:Company>
                            <efac:TouchPoint>
                                <cac:PartyIdentification>
                                    <cbc:ID schemeName="touchpoint">TPO-0001</cbc:ID>
                                </cac:PartyIdentification>
                            </efac:TouchPoint>
                        </efac:Organization>
                    </efac:Organizations>
                </efext:EformsExtension>
            </ext:ExtensionContent>
        </ext:UBLExtension>
    </ext:UBLExtensions>
    <cbc:UBLVersionID>2.3</cbc:UBLVersionID>
    <cbc:CustomizationID>eforms-sdk-1.10</cbc:CustomizationID>
    <cbc:ID schemeName="notice-id">a3a19810-3bf8-4860-8bb9-82eb3d0cd9ac</cbc:ID>
    <cbc:ContractFolderID>aff8eedb-7c00-4b87-87f9-30b33f946232</cbc:ContractFolderID>
    <cbc:IssueDate>2024-04-08Z</cbc:IssueDate>
    <cbc:IssueTime>12:33:01Z</cbc:IssueTime>
    <cbc:VersionID>00</cbc:VersionID>
    <cbc:RequestedPublicationDate>2024-04-08Z</cbc:RequestedPublicationDate>
    <cbc:RegulatoryDomain>31985R2137</cbc:RegulatoryDomain>
    <cbc:NoticeTypeCode listName="competition">qu-sy</cbc:NoticeTypeCode>
    <cbc:NoticeLanguageCode>POL</cbc:NoticeLanguageCode>
    <cac:ContractingParty>
        <cac:ContractingPartyType>
            <cbc:PartyTypeCode listName="buyer-legal-type">grp-p-aut</cbc:PartyTypeCode>
        </cac:ContractingPartyType>
        <cac:ContractingActivity>
            <cbc:ActivityTypeCode listName="authority-activity">hc-am</cbc:ActivityTypeCode>
        </cac:ContractingActivity>
        <cac:ContractingActivity>
            <cbc:ActivityTypeCode listName="entity-activity">electricity</cbc:ActivityTypeCode>
        </cac:ContractingActivity>
        <cac:Party>
            <cac:PartyIdentification>
                <cbc:ID schemeName="organization">ORG-0001</cbc:ID>
            </cac:PartyIdentification>
        </cac:Party>
    </cac:ContractingParty>
    <cac:ProcurementProject>
        <cbc:Name languageID="POL">example</cbc:Name>
        <cbc:Description languageID="POL">example</cbc:Description>
        <cbc:ProcurementTypeCode listName="contract-nature">services</cbc:ProcurementTypeCode>
        <cac:MainCommodityClassification>
            <cbc:ItemClassificationCode listName="cpv">98000000</cbc:ItemClassificationCode>
        </cac:MainCommodityClassification>
        <cac:RealizedLocation>
            <cbc:Description languageID="POL">example</cbc:Description>
            <cac:Address>
                <cbc:CountrySubentityCode listName="nuts">PL911</cbc:CountrySubentityCode>
                <cac:Country>
                    <cbc:IdentificationCode listName="country">POL</cbc:IdentificationCode>
                </cac:Country>
            </cac:Address>
        </cac:RealizedLocation>
    </cac:ProcurementProject>
    <cac:ProcurementProjectLot>
        <cbc:ID schemeName="Lot">LOT-0001</cbc:ID>
        <cac:TenderingTerms>
            <ext:UBLExtensions>
                <ext:UBLExtension>
                    <ext:ExtensionContent>
                        <efext:EformsExtension>
                            <efac:SelectionCriteria>
                                <cbc:CriterionTypeCode listName="selection-criterion">other</cbc:CriterionTypeCode>
                            </efac:SelectionCriteria>
                        </efext:EformsExtension>
                    </ext:ExtensionContent>
                </ext:UBLExtension>
            </ext:UBLExtensions>
            <cac:TendererQualificationRequest>
                <cac:SpecificTendererRequirement>
                    <cbc:TendererRequirementTypeCode listName="reserved-procurement">none</cbc:TendererRequirementTypeCode>
                </cac:SpecificTendererRequirement>
            </cac:TendererQualificationRequest>
            <cac:ContractExecutionRequirement>
                <cbc:ExecutionRequirementCode listName="reserved-execution">not-known</cbc:ExecutionRequirementCode>
            </cac:ContractExecutionRequirement>
            <cac:AdditionalInformationParty>
                <cac:PartyIdentification>
                    <cbc:ID schemeName="touchpoint">TPO-0001</cbc:ID>
                </cac:PartyIdentification>
            </cac:AdditionalInformationParty>
        </cac:TenderingTerms>
        <cac:TenderingProcess>
            <cbc:SubmissionMethodCode listName="esubmission">not-allowed</cbc:SubmissionMethodCode>
            <cbc:GovernmentAgreementConstraintIndicator>false</cbc:GovernmentAgreementConstraintIndicator>
            <cac:ProcessJustification>
                <cbc:ProcessReasonCode listName="no-esubmission-justification">ipr-iss</cbc:ProcessReasonCode>
                <cbc:Description languageID="POL">TEST</cbc:Description>
            </cac:ProcessJustification>
        </cac:TenderingProcess>
        <cac:ProcurementProject>
            <cbc:ID schemeName="InternalID">example1</cbc:ID>
            <cbc:Name languageID="POL">example2</cbc:Name>
            <cbc:Description languageID="POL">example</cbc:Description>
            <cbc:ProcurementTypeCode listName="contract-nature">services</cbc:ProcurementTypeCode>
            <cac:MainCommodityClassification>
                <cbc:ItemClassificationCode listName="cpv">66000000</cbc:ItemClassificationCode>
            </cac:MainCommodityClassification>
            <cac:RealizedLocation>
                <cbc:Description languageID="POL">desc</cbc:Description>
                <cac:Address>
                    <cbc:CountrySubentityCode listName="nuts">PL911</cbc:CountrySubentityCode>
                    <cac:Country>
                        <cbc:IdentificationCode listName="country">POL</cbc:IdentificationCode>
                    </cac:Country>
                </cac:Address>
            </cac:RealizedLocation>
            <cac:PlannedPeriod>
                <cbc:DescriptionCode listName="timeperiod">UNKNOWN</cbc:DescriptionCode>
            </cac:PlannedPeriod>
            <cac:ContractExtension>
                <cbc:MaximumNumberNumeric>1</cbc:MaximumNumberNumeric>
            </cac:ContractExtension>
        </cac:ProcurementProject>
    </cac:ProcurementProjectLot>
</ContractNotice>`

Thanks

**Reply by YvesJo** [reply]
Hi, 
Thanks. I actually meant an example of a field for which the condition wouldn't have any context

**Reply by Tronquill** [reply]
For field BT-5421-Lot in notice 15 I try to determine whether the field is required or not. 
When we try to validate the field BT-5421-Lot, and  it is not yet filled in, first  we  have to determine if BT-541-Lot-WeightNumber is not present. To do this, we have to take ND-LotAwardWeightCriterionParameter (context node). This node is not created though because user did not fill in BT-541-Lot-WeightNumber field (which is not required).

**Reply by YvesJo** [reply, marked as answer]
Thanks for this clarification, I thought it was about the absence of context in the condition EFX expression (which may not be the case) while it is about the context existence in the XML.
The existence of BT-5421-Lot is conditioned by the one of BT-541-Lot (forbidden when it does not exist and mandatory when it does).
The rule has been expressed one way only (BT-5421-Lot = conditioned by => BT-541-Lot) and not both ways as it would increase the number of rules. There are also cases were more than one fields are interdependent from a given fields and expressing rules in all possible directions would increase the number of rules exponentially causing performance issues and becoming a real challenge regarding their maintenance.
BT-5421 without its associated BT-541 is useless from a Business Perspective, and so does it for BT-541 without BT-5421. 
BT-541 is in the fact the "entry point" for this set of information and should be first filled in, it is optional because the set of information is itself optional.
I could not find any of the 2 fields in the XML shared above, BT-5421-Lot may have been removed at XML generation as without BT-541-Lot-WeightNumber, the information is useless.
There are various approaches to ensure information is filled in properly:

- clearly identify on the UI or draw the user attention to the fields that must first be filled in,
- have a "live validation" invoking rules relevant to a given group of information (display group) with the possibility to hide or display fields as they are forbidden/allowed

**Reply by Tronquill** [reply]
Thank you for your answer
